// speech-rule-engine/cjs/audio/auditory_description.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuditoryDescription = void 0;
const grammar_js_1 = require("../rule_engine/grammar.js");
const span_js_1 = require("./span.js");
class AuditoryDescription {
    static create(args, flags = {}) {
        args.text = grammar_js_1.Grammar.getInstance().apply(args.text, flags);
        return new AuditoryDescription(args);
    }
    constructor({ context, text, userValue, annotation, attributes, personality, layout }) {
        this.context = context || '';
        this.text = text || '';
        this.userValue = userValue || '';
        this.annotation = annotation || '';
        this.attributes = attributes || {};
        this.personality = personality || {};
        this.layout = layout || '';
    }
    isEmpty() {
        return (this.context.length === 0 &&
            this.text.length === 0 &&
            this.userValue.length === 0 &&
            this.annotation.length === 0);
    }
    clone() {
        let personality;
        if (this.personality) {
            personality = {};
            for (const [key, val] of Object.entries(this.personality)) {
                personality[key] = val;
            }
        }
        let attributes;
        if (this.attributes) {
            attributes = {};
            for (const [key, val] of Object.entries(this.attributes)) {
                attributes[key] = val;
            }
        }
        return new AuditoryDescription({
            context: this.context,
            text: this.text,
            userValue: this.userValue,
            annotation: this.annotation,
            personality: personality,
            attributes: attributes,
            layout: this.layout
        });
    }
    toString() {
        return ('AuditoryDescription(context="' +
            this.context +
            '" ' +
            ' text="' +
            this.text +
            '" ' +
            ' userValue="' +
            this.userValue +
            '" ' +
            ' annotation="' +
            this.annotation +
            '")');
    }
    descriptionString() {
        return this.context && this.text
            ? this.context + ' ' + this.text
            : this.context || this.text;
    }
    descriptionSpan() {
        return span_js_1.Span.stringAttr(this.descriptionString(), this.attributes);
    }
    equals(that) {
        return (this.context === that.context &&
            this.text === that.text &&
            this.userValue === that.userValue &&
            this.annotation === that.annotation);
    }
}
exports.AuditoryDescription = AuditoryDescription;
