// speech-rule-engine/cjs/audio/aural_rendering.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isXml = exports.registerRenderer = exports.error = exports.finalize = exports.merge = exports.markup = exports.getSeparator = exports.setSeparator = void 0;
const engine_js_1 = require("../common/engine.js");
const EngineConst = require("../common/engine_const.js");
const acss_renderer_js_1 = require("./acss_renderer.js");
const layout_renderer_js_1 = require("./layout_renderer.js");
const punctuation_renderer_js_1 = require("./punctuation_renderer.js");
const sable_renderer_js_1 = require("./sable_renderer.js");
const span_js_1 = require("./span.js");
const ssml_renderer_js_1 = require("./ssml_renderer.js");
const string_renderer_js_1 = require("./string_renderer.js");
const xml_renderer_js_1 = require("./xml_renderer.js");
const xmlInstance = new ssml_renderer_js_1.SsmlRenderer();
const renderers = new Map([
    [EngineConst.Markup.NONE, new string_renderer_js_1.StringRenderer()],
    [EngineConst.Markup.PUNCTUATION, new punctuation_renderer_js_1.PunctuationRenderer()],
    [EngineConst.Markup.LAYOUT, new layout_renderer_js_1.LayoutRenderer()],
    [EngineConst.Markup.ACSS, new acss_renderer_js_1.AcssRenderer()],
    [EngineConst.Markup.SABLE, new sable_renderer_js_1.SableRenderer()],
    [EngineConst.Markup.VOICEXML, xmlInstance],
    [EngineConst.Markup.SSML, xmlInstance]
]);
function setSeparator(sep) {
    const renderer = renderers.get(engine_js_1.default.getInstance().markup);
    if (renderer) {
        renderer.separator = sep;
    }
}
exports.setSeparator = setSeparator;
function getSeparator() {
    const renderer = renderers.get(engine_js_1.default.getInstance().markup);
    return renderer ? renderer.separator : '';
}
exports.getSeparator = getSeparator;
function markup(descrs) {
    const renderer = renderers.get(engine_js_1.default.getInstance().markup);
    if (!renderer) {
        return '';
    }
    return renderer.markup(descrs);
}
exports.markup = markup;
function merge(strs) {
    const span = strs.map((s) => {
        return typeof s === 'string' ? span_js_1.Span.stringEmpty(s) : s;
    });
    const renderer = renderers.get(engine_js_1.default.getInstance().markup);
    if (!renderer) {
        return strs.join();
    }
    return renderer.merge(span);
}
exports.merge = merge;
function finalize(str) {
    const renderer = renderers.get(engine_js_1.default.getInstance().markup);
    if (!renderer) {
        return str;
    }
    return renderer.finalize(str);
}
exports.finalize = finalize;
function error(key) {
    const renderer = renderers.get(engine_js_1.default.getInstance().markup);
    if (!renderer) {
        return '';
    }
    return renderer.error(key);
}
exports.error = error;
function registerRenderer(type, renderer) {
    renderers.set(type, renderer);
}
exports.registerRenderer = registerRenderer;
function isXml() {
    return renderers.get(engine_js_1.default.getInstance().markup) instanceof xml_renderer_js_1.XmlRenderer;
}
exports.isXml = isXml;
