// speech-rule-engine/cjs/speech_rules/mathspeak_french_util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.leftSubscriptBrief = exports.leftSuperscriptBrief = exports.leftSubscriptVerbose = exports.leftSuperscriptVerbose = exports.baselineBrief = exports.baselineVerbose = void 0;
const MathspeakUtil = require("./mathspeak_util.js");
function baselineVerbose(node) {
    const baseline = MathspeakUtil.baselineVerbose(node);
    baseline[0].speech = baseline[0].speech.replace(/-$/, '');
    return baseline;
}
exports.baselineVerbose = baselineVerbose;
function baselineBrief(node) {
    const baseline = MathspeakUtil.baselineBrief(node);
    baseline[0].speech = baseline[0].speech.replace(/-$/, '');
    return baseline;
}
exports.baselineBrief = baselineBrief;
function leftSuperscriptVerbose(node) {
    const leftIndex = MathspeakUtil.superscriptVerbose(node);
    leftIndex[0].speech = leftIndex[0].speech.replace(/^exposant/, 'exposant gauche');
    return leftIndex;
}
exports.leftSuperscriptVerbose = leftSuperscriptVerbose;
function leftSubscriptVerbose(node) {
    const leftIndex = MathspeakUtil.subscriptVerbose(node);
    leftIndex[0].speech = leftIndex[0].speech.replace(/^indice/, 'indice gauche');
    return leftIndex;
}
exports.leftSubscriptVerbose = leftSubscriptVerbose;
function leftSuperscriptBrief(node) {
    const leftIndex = MathspeakUtil.superscriptBrief(node);
    leftIndex[0].speech = leftIndex[0].speech.replace(/^sup/, 'sup gauche');
    return leftIndex;
}
exports.leftSuperscriptBrief = leftSuperscriptBrief;
function leftSubscriptBrief(node) {
    const leftIndex = MathspeakUtil.subscriptBrief(node);
    leftIndex[0].speech = leftIndex[0].speech.replace(/^sub/, 'sub gauche');
    return leftIndex;
}
exports.leftSubscriptBrief = leftSubscriptBrief;
