// speech-rule-engine/cjs/speech_rules/mathspeak_util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.smallRoot = exports.generateTensorRules = exports.removeParens = exports.generateBaselineConstraint = exports.determinantIsSimple = exports.nestedOverscript = exports.endscripts = exports.nestedUnderscript = exports.indexRadicalSbrief = exports.openingRadicalSbrief = exports.indexRadicalBrief = exports.closingRadicalBrief = exports.openingRadicalBrief = exports.indexRadicalVerbose = exports.closingRadicalVerbose = exports.openingRadicalVerbose = exports.baselineBrief = exports.baselineVerbose = exports.superscriptBrief = exports.superscriptVerbose = exports.subscriptBrief = exports.subscriptVerbose = exports.nestedSubSuper = exports.isSmallVulgarFraction = exports.overFractionSbrief = exports.closingFractionSbrief = exports.openingFractionSbrief = exports.closingFractionBrief = exports.openingFractionBrief = exports.overFractionVerbose = exports.closingFractionVerbose = exports.openingFractionVerbose = exports.fractionNestingDepth = exports.resetNestingDepth = exports.spaceoutIdentifier = exports.spaceoutNumber = exports.spaceoutText = void 0;
const span_js_1 = require("../audio/span.js");
const BaseUtil = require("../common/base_util.js");
const DomUtil = require("../common/dom_util.js");
const XpathUtil = require("../common/xpath_util.js");
const locale_js_1 = require("../l10n/locale.js");
const semantic_meaning_js_1 = require("../semantic_tree/semantic_meaning.js");
const semantic_processor_js_1 = require("../semantic_tree/semantic_processor.js");
let nestingDepth = {};
function spaceoutText(node) {
    return Array.from(node.textContent).map(span_js_1.Span.stringEmpty);
}
exports.spaceoutText = spaceoutText;
function spaceoutNodes(node, correction) {
    const content = Array.from(node.textContent);
    const result = [];
    const processor = semantic_processor_js_1.default.getInstance();
    const doc = node.ownerDocument;
    for (let i = 0, chr; (chr = content[i]); i++) {
        const leaf = processor
            .getNodeFactory()
            .makeLeafNode(chr, semantic_meaning_js_1.SemanticFont.UNKNOWN);
        const sn = processor.identifierNode(leaf, semantic_meaning_js_1.SemanticFont.UNKNOWN, '');
        correction(sn);
        result.push(sn.xml(doc));
    }
    return result;
}
function spaceoutNumber(node) {
    return spaceoutNodes(node, function (sn) {
        if (!sn.textContent.match(/\W/)) {
            sn.type = "number";
        }
    });
}
exports.spaceoutNumber = spaceoutNumber;
function spaceoutIdentifier(node) {
    return spaceoutNodes(node, function (sn) {
        sn.font = semantic_meaning_js_1.SemanticFont.UNKNOWN;
        sn.type = "identifier";
    });
}
exports.spaceoutIdentifier = spaceoutIdentifier;
const nestingBarriers = [
    "cases",
    "cell",
    "integral",
    "line",
    "matrix",
    "multiline",
    "overscore",
    "root",
    "row",
    "sqrt",
    "subscript",
    "superscript",
    "table",
    "underscore",
    "vector"
];
function resetNestingDepth(node) {
    nestingDepth = {};
    return [node];
}
exports.resetNestingDepth = resetNestingDepth;
function getNestingDepth(type, node, tags, opt_barrierTags, opt_barrierAttrs, opt_func) {
    opt_barrierTags = opt_barrierTags || nestingBarriers;
    opt_barrierAttrs = opt_barrierAttrs || {};
    opt_func =
        opt_func ||
            function (_node) {
                return false;
            };
    const xmlText = DomUtil.serializeXml(node);
    if (!nestingDepth[type]) {
        nestingDepth[type] = {};
    }
    if (nestingDepth[type][xmlText]) {
        return nestingDepth[type][xmlText];
    }
    if (opt_func(node) || tags.indexOf(node.tagName) < 0) {
        return 0;
    }
    const depth = computeNestingDepth_(node, tags, BaseUtil.setdifference(opt_barrierTags, tags), opt_barrierAttrs, opt_func, 0);
    nestingDepth[type][xmlText] = depth;
    return depth;
}
function containsAttr(node, attrs) {
    if (!node.attributes) {
        return false;
    }
    const attributes = DomUtil.toArray(node.attributes);
    for (let i = 0, attr; (attr = attributes[i]); i++) {
        if (attrs[attr.nodeName] === attr.nodeValue) {
            return true;
        }
    }
    return false;
}
function computeNestingDepth_(node, tags, barriers, attrs, func, depth) {
    if (func(node) ||
        barriers.indexOf(node.tagName) > -1 ||
        containsAttr(node, attrs)) {
        return depth;
    }
    if (tags.indexOf(node.tagName) > -1) {
        depth++;
    }
    if (!node.childNodes || node.childNodes.length === 0) {
        return depth;
    }
    const children = DomUtil.toArray(node.childNodes);
    return Math.max.apply(null, children.map(function (subNode) {
        return computeNestingDepth_(subNode, tags, barriers, attrs, func, depth);
    }));
}
function fractionNestingDepth(node) {
    return getNestingDepth('fraction', node, ['fraction'], nestingBarriers, {}, locale_js_1.LOCALE.FUNCTIONS.fracNestDepth);
}
exports.fractionNestingDepth = fractionNestingDepth;
function nestedFraction(node, expr, opt_end) {
    const depth = fractionNestingDepth(node);
    const annotation = Array(depth).fill(expr);
    if (opt_end) {
        annotation.push(opt_end);
    }
    return annotation.join(locale_js_1.LOCALE.MESSAGES.regexp.JOINER_FRAC);
}
function openingFractionVerbose(node) {
    return span_js_1.Span.singleton(nestedFraction(node, locale_js_1.LOCALE.MESSAGES.MS.START, locale_js_1.LOCALE.MESSAGES.MS.FRAC_V));
}
exports.openingFractionVerbose = openingFractionVerbose;
function closingFractionVerbose(node) {
    return span_js_1.Span.singleton(nestedFraction(node, locale_js_1.LOCALE.MESSAGES.MS.END, locale_js_1.LOCALE.MESSAGES.MS.FRAC_V), { kind: 'LAST' });
}
exports.closingFractionVerbose = closingFractionVerbose;
function overFractionVerbose(node) {
    return span_js_1.Span.singleton(nestedFraction(node, locale_js_1.LOCALE.MESSAGES.MS.FRAC_OVER), {});
}
exports.overFractionVerbose = overFractionVerbose;
function openingFractionBrief(node) {
    return span_js_1.Span.singleton(nestedFraction(node, locale_js_1.LOCALE.MESSAGES.MS.START, locale_js_1.LOCALE.MESSAGES.MS.FRAC_B));
}
exports.openingFractionBrief = openingFractionBrief;
function closingFractionBrief(node) {
    return span_js_1.Span.singleton(nestedFraction(node, locale_js_1.LOCALE.MESSAGES.MS.END, locale_js_1.LOCALE.MESSAGES.MS.FRAC_B), { kind: 'LAST' });
}
exports.closingFractionBrief = closingFractionBrief;
function openingFractionSbrief(node) {
    const depth = fractionNestingDepth(node);
    return span_js_1.Span.singleton(depth === 1
        ? locale_js_1.LOCALE.MESSAGES.MS.FRAC_S
        : locale_js_1.LOCALE.FUNCTIONS.combineNestedFraction(locale_js_1.LOCALE.MESSAGES.MS.NEST_FRAC, locale_js_1.LOCALE.FUNCTIONS.radicalNestDepth(depth - 1), locale_js_1.LOCALE.MESSAGES.MS.FRAC_S));
}
exports.openingFractionSbrief = openingFractionSbrief;
function closingFractionSbrief(node) {
    const depth = fractionNestingDepth(node);
    return span_js_1.Span.singleton(depth === 1
        ? locale_js_1.LOCALE.MESSAGES.MS.ENDFRAC
        : locale_js_1.LOCALE.FUNCTIONS.combineNestedFraction(locale_js_1.LOCALE.MESSAGES.MS.NEST_FRAC, locale_js_1.LOCALE.FUNCTIONS.radicalNestDepth(depth - 1), locale_js_1.LOCALE.MESSAGES.MS.ENDFRAC), { kind: 'LAST' });
}
exports.closingFractionSbrief = closingFractionSbrief;
function overFractionSbrief(node) {
    const depth = fractionNestingDepth(node);
    return span_js_1.Span.singleton(depth === 1
        ? locale_js_1.LOCALE.MESSAGES.MS.FRAC_OVER
        : locale_js_1.LOCALE.FUNCTIONS.combineNestedFraction(locale_js_1.LOCALE.MESSAGES.MS.NEST_FRAC, locale_js_1.LOCALE.FUNCTIONS.radicalNestDepth(depth - 1), locale_js_1.LOCALE.MESSAGES.MS.FRAC_OVER));
}
exports.overFractionSbrief = overFractionSbrief;
function isSmallVulgarFraction(node) {
    return locale_js_1.LOCALE.FUNCTIONS.fracNestDepth(node) ? [node] : [];
}
exports.isSmallVulgarFraction = isSmallVulgarFraction;
function nestedSubSuper(node, init, replace) {
    while (node.parentNode) {
        const children = node.parentNode;
        const parent = children.parentNode;
        if (!parent) {
            break;
        }
        const nodeRole = node.getAttribute && node.getAttribute('role');
        if ((parent.tagName === "subscript" &&
            node === children.childNodes[1]) ||
            (parent.tagName === "tensor" &&
                nodeRole &&
                (nodeRole === "leftsub" ||
                    nodeRole === "rightsub"))) {
            init = replace.sub + locale_js_1.LOCALE.MESSAGES.regexp.JOINER_SUBSUPER + init;
        }
        if ((parent.tagName === "superscript" &&
            node === children.childNodes[1]) ||
            (parent.tagName === "tensor" &&
                nodeRole &&
                (nodeRole === "leftsuper" ||
                    nodeRole === "rightsuper"))) {
            init = replace.sup + locale_js_1.LOCALE.MESSAGES.regexp.JOINER_SUBSUPER + init;
        }
        node = parent;
    }
    return init.trim();
}
exports.nestedSubSuper = nestedSubSuper;
function subscriptVerbose(node) {
    return span_js_1.Span.singleton(nestedSubSuper(node, locale_js_1.LOCALE.MESSAGES.MS.SUBSCRIPT, {
        sup: locale_js_1.LOCALE.MESSAGES.MS.SUPER,
        sub: locale_js_1.LOCALE.MESSAGES.MS.SUB
    }));
}
exports.subscriptVerbose = subscriptVerbose;
function subscriptBrief(node) {
    return span_js_1.Span.singleton(nestedSubSuper(node, locale_js_1.LOCALE.MESSAGES.MS.SUB, {
        sup: locale_js_1.LOCALE.MESSAGES.MS.SUP,
        sub: locale_js_1.LOCALE.MESSAGES.MS.SUB
    }));
}
exports.subscriptBrief = subscriptBrief;
function superscriptVerbose(node) {
    return span_js_1.Span.singleton(nestedSubSuper(node, locale_js_1.LOCALE.MESSAGES.MS.SUPERSCRIPT, {
        sup: locale_js_1.LOCALE.MESSAGES.MS.SUPER,
        sub: locale_js_1.LOCALE.MESSAGES.MS.SUB
    }));
}
exports.superscriptVerbose = superscriptVerbose;
function superscriptBrief(node) {
    return span_js_1.Span.singleton(nestedSubSuper(node, locale_js_1.LOCALE.MESSAGES.MS.SUP, {
        sup: locale_js_1.LOCALE.MESSAGES.MS.SUP,
        sub: locale_js_1.LOCALE.MESSAGES.MS.SUB
    }));
}
exports.superscriptBrief = superscriptBrief;
function baselineVerbose(node) {
    const baseline = nestedSubSuper(node, '', {
        sup: locale_js_1.LOCALE.MESSAGES.MS.SUPER,
        sub: locale_js_1.LOCALE.MESSAGES.MS.SUB
    });
    return span_js_1.Span.singleton(!baseline
        ? locale_js_1.LOCALE.MESSAGES.MS.BASELINE
        : baseline
            .replace(new RegExp(locale_js_1.LOCALE.MESSAGES.MS.SUB + '$'), locale_js_1.LOCALE.MESSAGES.MS.SUBSCRIPT)
            .replace(new RegExp(locale_js_1.LOCALE.MESSAGES.MS.SUPER + '$'), locale_js_1.LOCALE.MESSAGES.MS.SUPERSCRIPT));
}
exports.baselineVerbose = baselineVerbose;
function baselineBrief(node) {
    const baseline = nestedSubSuper(node, '', {
        sup: locale_js_1.LOCALE.MESSAGES.MS.SUP,
        sub: locale_js_1.LOCALE.MESSAGES.MS.SUB
    });
    return span_js_1.Span.singleton(baseline || locale_js_1.LOCALE.MESSAGES.MS.BASE);
}
exports.baselineBrief = baselineBrief;
function radicalNestingDepth(node) {
    return getNestingDepth('radical', node, ['sqrt', 'root'], nestingBarriers, {});
}
function nestedRadical(node, prefix, postfix) {
    const depth = radicalNestingDepth(node);
    const index = getRootIndex(node);
    postfix = index ? locale_js_1.LOCALE.FUNCTIONS.combineRootIndex(postfix, index) : postfix;
    return depth === 1
        ? postfix
        : locale_js_1.LOCALE.FUNCTIONS.combineNestedRadical(prefix, locale_js_1.LOCALE.FUNCTIONS.radicalNestDepth(depth - 1), postfix);
}
function getRootIndex(node) {
    const content = node.tagName === 'sqrt'
        ? '2'
        :
            XpathUtil.evalXPath('children/*[1]', node)[0].textContent.trim();
    return locale_js_1.LOCALE.MESSAGES.MSroots[content] || '';
}
function openingRadicalVerbose(node) {
    return span_js_1.Span.singleton(nestedRadical(node, locale_js_1.LOCALE.MESSAGES.MS.NESTED, locale_js_1.LOCALE.MESSAGES.MS.STARTROOT));
}
exports.openingRadicalVerbose = openingRadicalVerbose;
function closingRadicalVerbose(node) {
    return span_js_1.Span.singleton(nestedRadical(node, locale_js_1.LOCALE.MESSAGES.MS.NESTED, locale_js_1.LOCALE.MESSAGES.MS.ENDROOT));
}
exports.closingRadicalVerbose = closingRadicalVerbose;
function indexRadicalVerbose(node) {
    return span_js_1.Span.singleton(nestedRadical(node, locale_js_1.LOCALE.MESSAGES.MS.NESTED, locale_js_1.LOCALE.MESSAGES.MS.ROOTINDEX));
}
exports.indexRadicalVerbose = indexRadicalVerbose;
function openingRadicalBrief(node) {
    return span_js_1.Span.singleton(nestedRadical(node, locale_js_1.LOCALE.MESSAGES.MS.NEST_ROOT, locale_js_1.LOCALE.MESSAGES.MS.STARTROOT));
}
exports.openingRadicalBrief = openingRadicalBrief;
function closingRadicalBrief(node) {
    return span_js_1.Span.singleton(nestedRadical(node, locale_js_1.LOCALE.MESSAGES.MS.NEST_ROOT, locale_js_1.LOCALE.MESSAGES.MS.ENDROOT));
}
exports.closingRadicalBrief = closingRadicalBrief;
function indexRadicalBrief(node) {
    return span_js_1.Span.singleton(nestedRadical(node, locale_js_1.LOCALE.MESSAGES.MS.NEST_ROOT, locale_js_1.LOCALE.MESSAGES.MS.ROOTINDEX));
}
exports.indexRadicalBrief = indexRadicalBrief;
function openingRadicalSbrief(node) {
    return span_js_1.Span.singleton(nestedRadical(node, locale_js_1.LOCALE.MESSAGES.MS.NEST_ROOT, locale_js_1.LOCALE.MESSAGES.MS.ROOT));
}
exports.openingRadicalSbrief = openingRadicalSbrief;
function indexRadicalSbrief(node) {
    return span_js_1.Span.singleton(nestedRadical(node, locale_js_1.LOCALE.MESSAGES.MS.NEST_ROOT, locale_js_1.LOCALE.MESSAGES.MS.INDEX));
}
exports.indexRadicalSbrief = indexRadicalSbrief;
function underscoreNestingDepth(node) {
    return getNestingDepth('underscore', node, ['underscore'], nestingBarriers, {}, function (node) {
        return (node.tagName &&
            node.tagName === "underscore" &&
            node.childNodes[0].childNodes[1].getAttribute('role') ===
                "underaccent");
    });
}
function nestedUnderscript(node) {
    const depth = underscoreNestingDepth(node);
    return span_js_1.Span.singleton(Array(depth).join(locale_js_1.LOCALE.MESSAGES.MS.UNDER) + locale_js_1.LOCALE.MESSAGES.MS.UNDERSCRIPT);
}
exports.nestedUnderscript = nestedUnderscript;
function overscoreNestingDepth(node) {
    return getNestingDepth('overscore', node, ['overscore'], nestingBarriers, {}, function (node) {
        return (node.tagName &&
            node.tagName === "overscore" &&
            node.childNodes[0].childNodes[1].getAttribute('role') ===
                "overaccent");
    });
}
function endscripts(_node) {
    return span_js_1.Span.singleton(locale_js_1.LOCALE.MESSAGES.MS.ENDSCRIPTS);
}
exports.endscripts = endscripts;
function nestedOverscript(node) {
    const depth = overscoreNestingDepth(node);
    return span_js_1.Span.singleton(Array(depth).join(locale_js_1.LOCALE.MESSAGES.MS.OVER) + locale_js_1.LOCALE.MESSAGES.MS.OVERSCRIPT);
}
exports.nestedOverscript = nestedOverscript;
function determinantIsSimple(node) {
    if (node.tagName !== "matrix" ||
        node.getAttribute('role') !== "determinant") {
        return [];
    }
    const cells = XpathUtil.evalXPath('children/row/children/cell/children/*', node);
    for (let i = 0, cell; (cell = cells[i]); i++) {
        if (cell.tagName === "number") {
            continue;
        }
        if (cell.tagName === "identifier") {
            const role = cell.getAttribute('role');
            if (role === "latinletter" ||
                role === "greekletter" ||
                role === "otherletter") {
                continue;
            }
        }
        return [];
    }
    return [node];
}
exports.determinantIsSimple = determinantIsSimple;
function generateBaselineConstraint() {
    const ignoreElems = ['subscript', 'superscript', 'tensor'];
    const mainElems = ['relseq', 'multrel'];
    const breakElems = ['fraction', 'punctuation', 'fenced', 'sqrt', 'root'];
    const ancestrify = (elemList) => elemList.map((elem) => 'ancestor::' + elem);
    const notify = (elem) => 'not(' + elem + ')';
    const prefix = 'ancestor::*/following-sibling::*';
    const middle = notify(ancestrify(ignoreElems).join(' or '));
    const mainList = ancestrify(mainElems);
    const breakList = ancestrify(breakElems);
    let breakCstrs = [];
    for (let i = 0, brk; (brk = breakList[i]); i++) {
        breakCstrs = breakCstrs.concat(mainList.map(function (elem) {
            return brk + '/' + elem;
        }));
    }
    const postfix = notify(breakCstrs.join(' | '));
    return [[prefix, middle, postfix].join(' and ')];
}
exports.generateBaselineConstraint = generateBaselineConstraint;
function removeParens(node) {
    if (!node.childNodes.length ||
        !node.childNodes[0].childNodes.length ||
        !node.childNodes[0].childNodes[0].childNodes.length) {
        return span_js_1.Span.singleton('');
    }
    const content = node.childNodes[0].childNodes[0].childNodes[0].textContent;
    return span_js_1.Span.singleton(content.match(/^\(.+\)$/) ? content.slice(1, -1) : content);
}
exports.removeParens = removeParens;
const componentString = new Map([
    [3, 'CSFleftsuperscript'],
    [4, 'CSFleftsubscript'],
    [2, 'CSFbaseline'],
    [1, 'CSFrightsubscript'],
    [0, 'CSFrightsuperscript']
]);
const childNumber = new Map([
    [4, 2],
    [3, 3],
    [2, 1],
    [1, 4],
    [0, 5]
]);
function generateTensorRuleStrings_(constellation) {
    const constraints = [];
    let verbString = '';
    let briefString = '';
    let constel = parseInt(constellation, 2);
    for (let i = 0; i < 5; i++) {
        const childString = 'children/*[' + childNumber.get(i) + ']';
        if (constel & 1) {
            const compString = componentString.get(i % 5);
            verbString =
                '[t] ' + compString + 'Verbose; [n] ' + childString + ';' + verbString;
            briefString =
                '[t] ' + compString + 'Brief; [n] ' + childString + ';' + briefString;
        }
        else {
            constraints.unshift('name(' + childString + ')="empty"');
        }
        constel >>= 1;
    }
    return [constraints, verbString, briefString];
}
function generateTensorRules(store, brief = true) {
    const constellations = [
        '11111',
        '11110',
        '11101',
        '11100',
        '10111',
        '10110',
        '10101',
        '10100',
        '01111',
        '01110',
        '01101',
        '01100'
    ];
    for (const constel of constellations) {
        let name = 'tensor' + constel;
        let [components, verbStr, briefStr] = generateTensorRuleStrings_(constel);
        store.defineRule(name, 'default', verbStr, 'self::tensor', ...components);
        if (brief) {
            store.defineRule(name, 'brief', briefStr, 'self::tensor', ...components);
            store.defineRule(name, 'sbrief', briefStr, 'self::tensor', ...components);
        }
        if (!(parseInt(constel, 2) & 3)) {
            continue;
        }
        const baselineStr = componentString.get(2);
        verbStr += '; [t]' + baselineStr + 'Verbose';
        briefStr += '; [t]' + baselineStr + 'Brief';
        name = name + '-baseline';
        const cstr = '((.//*[not(*)])[last()]/@id)!=(((.//ancestor::fraction|' +
            'ancestor::root|ancestor::sqrt|ancestor::cell|ancestor::line|' +
            'ancestor::stree)[1]//*[not(*)])[last()]/@id)';
        store.defineRule(name, 'default', verbStr, 'self::tensor', cstr, ...components);
        if (brief) {
            store.defineRule(name, 'brief', briefStr, 'self::tensor', cstr, ...components);
            store.defineRule(name, 'sbrief', briefStr, 'self::tensor', cstr, ...components);
        }
    }
}
exports.generateTensorRules = generateTensorRules;
function smallRoot(node) {
    let max = Object.keys(locale_js_1.LOCALE.MESSAGES.MSroots).length;
    if (!max) {
        return [];
    }
    else {
        max++;
    }
    if (!node.childNodes ||
        node.childNodes.length === 0 ||
        !node.childNodes[0].childNodes) {
        return [];
    }
    const index = node.childNodes[0].childNodes[0].textContent;
    if (!/^\d+$/.test(index)) {
        return [];
    }
    const num = parseInt(index, 10);
    return num > 1 && num <= max ? [node] : [];
}
exports.smallRoot = smallRoot;
